python/pyiceberg/table/__init__.py

```diff
@@ -355,7 +355,23 @@ def to_arrow(self):
         if "*" not in self.selected_fields:
             columns = list(self.selected_fields)
 
-        return pq.read_table(source=locations, filesystem=fs, columns=columns)
+        pyarrow_filter = None
+        if self.row_filter is not AlwaysTrue():
+            bound_row_filter = bind(self.table.schema(), self.row_filter)
+            pyarrow_filter = expression_to_pyarrow(bound_row_filter)
+
+        from pyarrow.dataset import dataset
+
+        ds = dataset(
+            source=locations,
+            filesystem=fs,
+            # Optionally provide the Schema for the Dataset,
+            # in which case it will not be inferred from the source.
+            # https://arrow.apache.org/docs/python/generated/pyarrow.dataset.dataset.html#pyarrow.dataset.dataset
+            schema=schema_to_pyarrow(self.table.schema()),
```

Comment:
Should this schema match the file schema? We should test a couple of projection cases.

Reply:
It should be equal or a subset of the original schema, see the PR description

Reply:
What I mean is that this probably needs to use the field names from each Parquet data file, not the field names from the table schema. That's because Parquet typically projects columns by name, but Iceberg uses IDs. There's no guarantee that the file's column names match the table's column names.